vshard/replicaset.lua

```diff
@@ -144,6 +144,70 @@ local function netbox_wait_connected(conn, timeout)
     return timeout
 end
 
+local netbox_is_conn_dead
+
+if not (util.version_is_at_least(2, 10, 0, 'beta', 2, 0) and
+        util.version_is_at_most(2, 10, 0, nil, 0, 0)) then
+
+--
+-- Check if the connection is dead and it won't be restored automatically.
+-- Even though replicaset's connections are initialized with the option
+-- 'reconnect_after', there's cases, when no reconnect will be done (e.g.
+-- if the conn was explicitly cancelled or the connection's worker fiber
+-- was killed). In these situations we need to reestablish it manually as
+-- the missing connection to some replicaset is a critical problem which
+-- leads to the inability of the user to access some part of a data via
+-- router.
+--
+-- Note: the function expects the conn to be non-nil.
+--
+netbox_is_conn_dead = function(conn)
+    assert(conn)
+    if conn.state == 'error' or conn.state == 'closed' then
+        return true
+    end
+    -- We need to check the 'error_reconnect' option more precisely and not to filter
+    -- it out right here because in some versions of tarantool in case of a killed worker
+    -- fiber no reconnection is done.
+    if conn.state ~= 'error_reconnect' then
+        -- The connection is fetching schema or doing auth, which
+        -- means it is not dead and shoudn't be reinitialized.
+        return false
+    end
+    if not conn._fiber then
+        -- The fiber field is not present in old netbox, which correctly reports
+        -- state as 'error' when its fiber got killed. It would be filtered out
+        -- above then.
+        return false
+    end
+    return conn._fiber:status() ~= "dead"
+end
+
+else -- 2.10.0
+
+--
+-- Tarantool 2.10.0 has a bug: after worker fiber kill connection's state is
+-- error_reconnect, but it doesn't reconnet anymore. The only way we can
+-- catch this situation is error string comparison as the connection doesn't
+-- have _fiber in itself.
+--
+netbox_is_conn_dead = function(conn)
+    assert(conn)
+    if conn.state == 'error' or conn.state == 'closed' then
+        return true
+    end
+
+    if conn.state ~= 'error_reconnect' then
+        -- The connection is fetching schema or doing auth, which
+        -- means it is not dead and shoudn't be reinitialized.
+        return false
+    end
+
+    return conn.error == 'fiber is cancelled'
```

Comment:
AFAICS, this is the only way we can filter out dead connection on buggy versions. It can still be triggered false-positively, but it seems to be the better option rather than ignoring these versions.